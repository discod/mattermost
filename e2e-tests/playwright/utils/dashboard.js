// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/* eslint-disable no-console */

const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs');
const path = require('path');
const {stringify} = require('querystring');

require('dotenv').config();

const {AUTOMATION_DASHBOARD_TOKEN, AUTOMATION_DASHBOARD_URL} = process.env;

const config = {
    baseURL: AUTOMATION_DASHBOARD_URL,
    headers: {
        Authorization: `Bearer ${AUTOMATION_DASHBOARD_TOKEN}`,
    },
};

/**
 * Create and start a test cycle.
 * @param {Object} data - Contains repo, branch, build and files.
 * @returns {Object} - Contains cycle information.
 */
async function createAndStartCycle(data) {
    try {
        console.log('Creating test cycle with dashboard at:', config.baseURL);

        const response = await axios.post('/api/cycles/start', data, config);
        return response.data;
    } catch (err) {
        console.error('Error creating test cycle:', err.message);
        if (err.response) {
            console.error('Status:', err.response.status);
            console.error('Data:', err.response.data);
        }
        throw err;
    }
}

/**
 * Get a spec to test.
 * @param {Object} data - Contains cycle_id.
 * @returns {Object} - Contains spec information.
 */
async function getSpecToTest(data) {
    try {
        // Use cycle_id to get specs
        const cycleId = data.cycle_id;
        if (!cycleId) {
            return {
                code: 'MISSING_CYCLE_ID',
                message: 'Cycle ID is required',
            };
        }

        console.log(`Getting specs for cycle: ${cycleId}`);
        const response = await axios.get(`/api/specs/to-test?cycle_id=${cycleId}`, config);

        if (!response.data || !Array.isArray(response.data.specs)) {
            return {
                code: 'INVALID_RESPONSE',
                message: 'Invalid response from dashboard API',
                data: response.data,
            };
        }

        // Return the specs array
        return {
            specs: response.data.specs,
            cycle: response.data.cycle,
        };
    } catch (err) {
        return {
            code: err.code || 'ERROR',
            message: err.message,
            data: err.response ? err.response.data : undefined,
        };
    }
}

/**
 * Record a spec result.
 * @param {string} id - Execution ID.
 * @param {Object} spec - Spec information.
 * @param {Array} testCases - Test cases information.
 * @returns {Object} - Contains updated spec information.
 */
async function recordSpecResult(id, spec, testCases) {
    try {
        const response = await axios.post(`/api/specs/${id}/result`, {spec, testCases}, config);
        return response.data;
    } catch (err) {
        console.error('Error recording spec result:', err.message);
        return {error: err.message};
    }
}

/**
 * Update a cycle.
 * @param {string} id - Cycle ID.
 * @param {Object} data - Data to update.
 * @returns {Object} - Contains updated cycle information.
 */
async function updateCycle(id, data) {
    try {
        const response = await axios.put(`/api/cycles/${id}`, data, config);
        return response.data;
    } catch (err) {
        console.error('Error updating cycle:', err.message);
        return {error: err.message};
    }
}

/**
 * Upload a screenshot.
 * @param {string} filePath - Path of the screenshot.
 * @param {string} repo - Repository name.
 * @param {string} branch - Branch name.
 * @param {string} build - Build identifier.
 * @returns {string} - URL of the uploaded screenshot.
 */
async function uploadScreenshot(filePath, repo, branch, build) {
    try {
        if (!fs.existsSync(filePath)) {
            return {error: `File does not exist: ${filePath}`};
        }

        const form = new FormData();
        form.append('repo', repo);
        form.append('branch', branch);
        form.append('build', build);
        form.append('file', fs.createReadStream(filePath));

        const response = await axios.post('/api/screenshots', form, {
            ...config,
            headers: {
                ...config.headers,
                ...form.getHeaders(),
            },
        });

        return response.data.url;
    } catch (err) {
        console.error('Error uploading screenshot:', err.message);
        return {error: err.message};
    }
}

module.exports = {
    createAndStartCycle,
    getSpecToTest,
    recordSpecResult,
    updateCycle,
    uploadScreenshot,
};
