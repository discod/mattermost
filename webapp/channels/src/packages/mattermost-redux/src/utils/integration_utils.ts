// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {DialogElement} from '@mattermost/types/integrations';

type DialogError = {
    id: string;
    defaultMessage: string;
    values?: any;
};
export function checkDialogElementForError(elem: DialogElement, value: any): DialogError | undefined | null {
    if ((!value && value !== 0) && !elem.optional) {
        return {
            id: 'interactive_dialog.error.required',
            defaultMessage: 'This field is required.',
        };
    }

    const type = elem.type;

    if (type === 'text' || type === 'textarea') {
        if (value && value.length < elem.min_length) {
            return {
                id: 'interactive_dialog.error.too_short',
                defaultMessage: 'Minimum input length is {minLength}.',
                values: {minLength: elem.min_length},
            };
        }

        if (elem.subtype === 'email') {
            if (value && !value.includes('@')) {
                return {
                    id: 'interactive_dialog.error.bad_email',
                    defaultMessage: 'Must be a valid email address.',
                };
            }
        }

        if (elem.subtype === 'number') {
            if (value && isNaN(value)) {
                return {
                    id: 'interactive_dialog.error.bad_number',
                    defaultMessage: 'Must be a number.',
                };
            }
        }

        if (elem.subtype === 'url') {
            if (value && !value.includes('http://') && !value.includes('https://')) {
                return {
                    id: 'interactive_dialog.error.bad_url',
                    defaultMessage: 'URL must include http:// or https://.',
                };
            }
        }
    } else if (type === 'radio') {
        const options = elem.options;

        if (typeof value !== 'undefined' && Array.isArray(options) && !options.some((e) => e.value === value)) {
            return {
                id: 'interactive_dialog.error.invalid_option',
                defaultMessage: 'Must be a valid option',
            };
        }
    } else if (type === 'date' || type === 'datetime') {
        // Import date utilities for validation
        const {stringToMoment, resolveRelativeDate} = require('utils/date_utils');
        
        // Validate min_date format if present
        if (elem.min_date) {
            const minDateResolved = resolveRelativeDate(elem.min_date);
            const minDateMoment = stringToMoment(minDateResolved);
            if (!minDateMoment) {
                return {
                    id: 'interactive_dialog.error.invalid_min_date',
                    defaultMessage: 'Invalid min_date format',
                };
            }
        }
        
        // Validate max_date format if present
        if (elem.max_date) {
            const maxDateResolved = resolveRelativeDate(elem.max_date);
            const maxDateMoment = stringToMoment(maxDateResolved);
            if (!maxDateMoment) {
                return {
                    id: 'interactive_dialog.error.invalid_max_date',
                    defaultMessage: 'Invalid max_date format',
                };
            }
        }
        
        if (value) {
            const date = stringToMoment(value);
            if (!date) {
                return {
                    id: 'interactive_dialog.error.invalid_date',
                    defaultMessage: 'Invalid date format',
                };
            }

            // Check min_date constraint
            if (elem.min_date) {
                const min = stringToMoment(resolveRelativeDate(elem.min_date));
                if (min && date.isBefore(min, 'day')) {
                    return {
                        id: 'interactive_dialog.error.date_too_early',
                        defaultMessage: 'Date must be after {minDate}',
                        values: {minDate: min.format('YYYY-MM-DD')},
                    };
                }
            }

            // Check max_date constraint
            if (elem.max_date) {
                const max = stringToMoment(resolveRelativeDate(elem.max_date));
                if (max && date.isAfter(max, 'day')) {
                    return {
                        id: 'interactive_dialog.error.date_too_late',
                        defaultMessage: 'Date must be before {maxDate}',
                        values: {maxDate: max.format('YYYY-MM-DD')},
                    };
                }
            }
        }
    }

    return null;
}

// If we're returned errors that don't match any of the elements we have,
// ignore them and complete the dialog

export function checkIfErrorsMatchElements(errors: Record<string, string> = {}, elements: DialogElement[] = []) {
    for (const name in errors) {
        if (!Object.hasOwn(errors, name)) {
            continue;
        }
        for (const elem of elements) {
            if (elem.name === name) {
                return true;
            }
        }
    }

    return false;
}
