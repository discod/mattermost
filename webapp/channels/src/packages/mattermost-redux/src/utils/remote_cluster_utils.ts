// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {DateTime, Interval} from 'luxon';

import type {RemoteCluster} from '@mattermost/types/remote_clusters';

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));
