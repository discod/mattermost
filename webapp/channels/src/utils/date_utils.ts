// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {parseISO, isValid} from 'date-fns';
import moment, {type Moment} from 'moment-timezone';
import {useMemo} from 'react';

export enum DateReference {

    // Absolute
    TODAY = 'today',
    TOMORROW = 'tomorrow',
    YESTERDAY = 'yesterday',
}

// Time validation regex for time-only validation
// HH:MM format (24-hour notation)
export const TIME_FORMAT_REGEX = /^([01]\d|2[0-3]):([0-5]\d)$/;

// RFC3339 datetime format with seconds for plugin compatibility
export const RFC3339_DATETIME_FORMAT = 'YYYY-MM-DDTHH:mm:ss[Z]';

/**
 * Convert a string value (ISO format or relative) to a Moment object
 * For date-only fields, datetime formats are accepted and the date portion is extracted
 */
export function stringToMoment(value: string | null, timezone?: string): Moment | null {
    if (!value) {
        return null;
    }

    // Handle relative dates/times - returns moment object directly to avoid double conversion
    const relativeMoment = resolveRelativeDateToMoment(value, timezone);
    if (relativeMoment) {
        return relativeMoment;
    }

    // Use parseISO for validation, but keep timezone logic separate
    let momentValue: moment.Moment;

    // First validate with parseISO
    try {
        const parsedDate = parseISO(value);
        if (!isValid(parsedDate)) {
            return null; // parseISO validation failed
        }
    } catch (error) {
        return null; // parseISO parsing failed
    }

    // parseISO validation passed, now use moment for timezone handling
    if (timezone && moment.tz.zone(timezone)) {
        momentValue = moment.tz(value, timezone);
    } else {
        momentValue = moment(value);
    }

    return momentValue.isValid() ? momentValue : null;
}

/**
 * Convert a Moment object to an ISO string for storage
 */
export function momentToString(momentValue: Moment | null, isDateTime: boolean): string | null {
    if (!momentValue || !momentValue.isValid()) {
        return null;
    }

    if (isDateTime) {
        // Store datetime in UTC format with seconds: "2025-01-14T14:30:00Z"
        return momentValue.utc().format(RFC3339_DATETIME_FORMAT);
    }

    // Store date only: "2025-01-14"
    return momentValue.format('YYYY-MM-DD');
}

/**
 * Resolve relative date references to Moment objects (internal helper)
 */
function resolveRelativeDateToMoment(dateStr: string, timezone?: string): Moment | null {
    const now = timezone && moment.tz.zone(timezone) ? moment.tz(timezone) : moment();

    switch (dateStr) {
    case DateReference.TODAY:
        return now.startOf('day');

    case DateReference.TOMORROW:
        return now.add(1, 'day').startOf('day');

    case DateReference.YESTERDAY:
        return now.subtract(1, 'day').startOf('day');

    default: {
        // Handle dynamic patterns like "+5d", "+2w", "+1m" (limit to 4 digits for security)
        const dynamicMatch = dateStr.match(/^([+-]\d{1,4})([dwm])$/);
        if (dynamicMatch) {
            const [, amount, unit] = dynamicMatch;
            const value = parseInt(amount, 10);

            // Additional bounds checking for security
            if (Math.abs(value) > 9999) {
                return null; // Return null if value is too large
            }

            let momentUnit: moment.unitOfTime.DurationConstructor;

            switch (unit) {
            case 'd':
                momentUnit = 'day';
                return now.add(value, momentUnit).startOf('day');
            case 'w':
                momentUnit = 'week';
                return now.add(value, momentUnit).startOf('day');
            case 'm':
                momentUnit = 'month';
                return now.add(value, momentUnit).startOf('day');
            default:
                return null; // Return null if pattern not recognized
            }
        }

        // Return null if not a recognized relative reference
        return null;
    }
    }
}

/**
 * Resolve relative date references to ISO strings
 */
export function resolveRelativeDate(dateStr: string, timezone?: string): string {
    // Try to resolve as relative date/time
    const relativeMoment = resolveRelativeDateToMoment(dateStr, timezone);
    if (relativeMoment) {
        // All relative patterns (d/w/m) return date format since we removed hour support
        return relativeMoment.format('YYYY-MM-DD');
    }

    // Return as-is if not a recognized relative reference
    return dateStr;
}

/**
 * Hook to memoize relative date resolution based on timezone and current time
 */
export function useMemoizedRelativeDate(dateStr: string, timezone?: string): string {
    const currentMinute = Math.floor(Date.now() / (1000 * 60));

    return useMemo(() => {
        return resolveRelativeDate(dateStr, timezone);
    }, [dateStr, timezone, currentMinute]); // Re-compute every minute
}

export interface DateValidationError {
    id: string;
    defaultMessage: string;
    values?: Record<string, string>;
}

/**
 * Validate if a date string is within min/max constraints
 * Returns structured validation error for internationalization
 */
export function validateDateRange(
    dateStr: string | null,
    minDate?: string,
    maxDate?: string,
    timezone?: string,
    locale?: string,
): DateValidationError | null {
    if (!dateStr) {
        return null;
    }

    const date = stringToMoment(dateStr, timezone);
    if (!date) {
        return {
            id: 'apps_form.date_field.invalid_format',
            defaultMessage: 'Invalid date format',
        };
    }

    if (minDate) {
        const min = stringToMoment(resolveRelativeDate(minDate, timezone), timezone);
        if (min && date.isBefore(min, 'day')) {
            // Format date for user preferences using Intl.DateTimeFormat
            const formattedDate = new Intl.DateTimeFormat(locale, {
                year: 'numeric',
                month: 'short',
                day: 'numeric',
            }).format(new Date(min.year(), min.month(), min.date()));

            return {
                id: 'apps_form.date_field.min_date_error',
                defaultMessage: 'Date must be after {minDate}',
                values: {minDate: formattedDate},
            };
        }
    }

    if (maxDate) {
        const max = stringToMoment(resolveRelativeDate(maxDate, timezone), timezone);
        if (max && date.isAfter(max, 'day')) {
            // Format date for user preferences using Intl.DateTimeFormat
            const formattedDate = new Intl.DateTimeFormat(locale, {
                year: 'numeric',
                month: 'short',
                day: 'numeric',
            }).format(new Date(max.year(), max.month(), max.date()));

            return {
                id: 'apps_form.date_field.max_date_error',
                defaultMessage: 'Date must be before {maxDate}',
                values: {maxDate: formattedDate},
            };
        }
    }

    return null;
}

