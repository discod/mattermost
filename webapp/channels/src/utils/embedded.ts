// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/**
 * Checks if the application is running embedded in another platform
 * This is determined by the MMEMBED cookie being set to '1'
 * @returns True if the app is running embedded
 */
export function isEmbedded(): boolean {
    try {
        return window.self !== window.parent;
    } catch (e) {
        // If accessing window.parent throws an error, we're in a cross-origin iframe
        return true;
    }
}
