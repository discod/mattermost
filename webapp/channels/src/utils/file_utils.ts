// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ClientConfig} from '@mattermost/types/config';
import type {FileInfo} from '@mattermost/types/files';

import {getFileDownloadUrl} from 'mattermost-redux/utils/file_utils';

import {isMobileApp} from 'utils/user_agent';

export const FileSizes = {

    // Bytes
    Byte: 1,
    Kilobyte: 1024,
    Megabyte: 1024 * 1024,
    Gigabyte: 1024 * 1024 * 1024,
    Terabyte: 1024 * 1024 * 1024 * 1024,
} as const;

export function isFileAttachmentsEnabled(config: Partial<ClientConfig>): boolean {
    return config.EnableFileAttachments === 'true';
}

export function canUploadFiles(config: Partial<ClientConfig>): boolean {
    const enableFileAttachments = isFileAttachmentsEnabled(config);
    const enableMobileFileUpload = isMobileFileUploadsEnabled(config);

    if (!enableFileAttachments) {
        return false;
    }

    if (isMobileApp()) {
        return enableMobileFileUpload;
    }

    return true;
}

export function isMobileFileUploadsEnabled(config: Partial<ClientConfig>): boolean {
    return config.EnableMobileFileUpload === 'true';
}

export function canDownloadFiles(config: Partial<ClientConfig>): boolean {
    if (isMobileApp()) {
        return config.EnableMobileFileDownload === 'true';
    }

    return true;
}

export function downloadMultipleFiles(files: FileInfo[]) {
    // Create a hidden container for the download links
    const container = document.createElement('div');
    container.style.display = 'none';
    document.body.appendChild(container);

    try {
        // Create and click download links for each file
        for (const file of files) {
            const link = document.createElement('a');
            link.href = getFileDownloadUrl(file.id);
            link.download = file.name;
            link.style.display = 'none';
            container.appendChild(link);
            link.click();
        }
    } finally {
        // Clean up the container
        document.body.removeChild(container);
    }
}
