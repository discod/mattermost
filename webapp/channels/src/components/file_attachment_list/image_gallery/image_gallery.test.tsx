import React from 'react';
import {render, screen, fireEvent, act} from '@testing-library/react';
import '@testing-library/jest-dom';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import ImageGallery from './image_gallery';
import {TestHelper} from 'utils/test_helper';
import {IntlProvider} from 'react-intl';

const mockStore = configureStore([]);
const store = mockStore({
    entities: {
        files: {
            files: {
                img1: {
                    id: 'img1',
                    name: 'image1.png',
                    extension: 'png',
                    width: 100,
                    height: 100,
                    link: 'http://example.com/image1.png',
                },
                img2: {
                    id: 'img2',
                    name: 'image2.jpg',
                    extension: 'jpg',
                    width: 300,
                    height: 200,
                    link: 'http://example.com/image2.jpg',
                },
            },
            fileIdsByPostId: {
                post1: ['img1', 'img2'],
            },
        },
        i18n: {
            locale: 'en',
        },
        general: {
            config: {},
        },
    },
    views: {
        rhs: {
            selectedPostId: '',
        },
    },
});

const mockFileInfos = [
    TestHelper.getFileInfoMock({
        id: 'img1',
        name: 'image1.png',
        extension: 'png',
        width: 100,
        height: 100,
        link: 'http://example.com/image1.png',
    }),
    TestHelper.getFileInfoMock({
        id: 'img2',
        name: 'image2.jpg',
        extension: 'jpg',
        width: 300,
        height: 200,
        link: 'http://example.com/image2.jpg',
    }),
];

const defaultProps = {
    fileInfos: mockFileInfos,
    canDownloadFiles: true,
    handleImageClick: jest.fn(),
    onToggleCollapse: jest.fn(),
    isEmbedVisible: true,
    postId: 'post1',
    allFilesForPost: mockFileInfos,
};

function renderWithProvider(ui) {
    return render(
        <Provider store={store}>
            <IntlProvider locale="en" messages={{}}>
                {ui}
            </IntlProvider>
        </Provider>
    );
}

describe('ImageGallery', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders with images', () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        expect(screen.getByTestId('fileAttachmentList')).toBeInTheDocument();
        expect(screen.getAllByRole('listitem')).toHaveLength(2);
    });

    it('renders with no images', () => {
        renderWithProvider(<ImageGallery {...defaultProps} fileInfos={[]} allFilesForPost={[]} />);
        expect(screen.getByTestId('fileAttachmentList')).toBeInTheDocument();
        expect(screen.queryAllByRole('listitem')).toHaveLength(0);
    });

    it('renders collapsed state', () => {
        renderWithProvider(<ImageGallery {...defaultProps} isEmbedVisible={false} />);
        expect(screen.getByRole('list')).toHaveClass('collapsed');
    });

    it('calls handleImageClick on Enter/Space', () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const items = screen.getAllByRole('listitem');
        fireEvent.keyDown(items[0], {key: 'Enter'});
        expect(defaultProps.handleImageClick).toHaveBeenCalled();
        fireEvent.keyDown(items[1], {key: ' '});
        expect(defaultProps.handleImageClick).toHaveBeenCalledTimes(2);
    });

    it('calls onToggleCollapse when toggled', () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const toggleBtn = screen.getByRole('button', {name: /images/i});
        fireEvent.click(toggleBtn);
        expect(defaultProps.onToggleCollapse).toHaveBeenCalled();
    });

    it('disables download button when downloading', async () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const downloadBtn = screen.getByRole('button', {name: /download all/i});
        await act(async () => {
            fireEvent.click(downloadBtn);
        });
        expect(downloadBtn).toBeDisabled();
    });

    it('disables download button when not allowed', () => {
        renderWithProvider(<ImageGallery {...defaultProps} canDownloadFiles={false} />);
        const downloadBtn = screen.getByRole('button', {name: /download all/i});
        expect(downloadBtn).toBeDisabled();
    });

    it('shows correct aria attributes', () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const toggleBtn = screen.getByRole('button', {name: /images/i});
        expect(toggleBtn).toHaveAttribute('aria-expanded');
        expect(screen.getByRole('list')).toBeInTheDocument();
    });

    it('triggers download for each file with correct filename and url', async () => {
        // Suppress jsdom navigation error for this test only
        const originalError = console.error;
        console.error = (msg, ...args) => {
            if (msg && msg.toString().includes('Not implemented: navigation')) {
                return;
            }
            originalError(msg, ...args);
        };
        
        // Spy on document.createElement and check the anchor's properties
        const aStub = document.createElement('a');
        aStub.click = jest.fn();
        const createElementSpy = jest.spyOn(document, 'createElement').mockImplementation((tagName) => {
            if (tagName === 'a') {
                return aStub;
            }
            return document.createElementNS('http://www.w3.org/1999/xhtml', tagName);
        });
        const appendChildSpy = jest.spyOn(document.body, 'appendChild');
        const removeChildSpy = jest.spyOn(document.body, 'removeChild');
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const downloadBtn = screen.getByRole('button', {name: /download all/i});
        await act(async () => {
            fireEvent.click(downloadBtn);
        });
        
        // Check that the download handler logic was triggered for each file
        expect(createElementSpy.mock.calls.filter(call => call[0] === 'a').length).toBe(defaultProps.fileInfos.length);
        expect(aStub.click).toHaveBeenCalledTimes(defaultProps.fileInfos.length);
        expect(appendChildSpy).toHaveBeenCalled();
        expect(removeChildSpy).toHaveBeenCalled();
        // Optionally, check that the anchor's download attribute is set to the sanitized filename
        expect(aStub.download).toBe('image2.jpg'); // last file in mockFileInfos
        expect(aStub.href).toBe('http://example.com/image2.jpg'); // last file in mockFileInfos
        console.error = originalError;
    });

    it('handles edge case: invalid fileInfo.link', async () => {
        const badFiles = [
            TestHelper.getFileInfoMock({id: 'bad', name: 'bad.png', extension: 'png', width: 100, height: 100, link: null}),
        ];
        const createElementSpy = jest.spyOn(document, 'createElement');
        renderWithProvider(<ImageGallery {...defaultProps} fileInfos={badFiles} allFilesForPost={badFiles} />);
        const downloadBtn = screen.getByRole('button', {name: /download all/i});
        await act(async () => {
            fireEvent.click(downloadBtn);
        });
        // Should be disabled for invalid link
        expect(downloadBtn).toBeDisabled();
        expect(createElementSpy).not.toHaveBeenCalledWith('a');
        createElementSpy.mockRestore();
    });

    it('updates aria-live region on collapse/expand', () => {
        renderWithProvider(<ImageGallery {...defaultProps} />);
        const toggleBtn = screen.getByRole('button', {name: /images/i});
        fireEvent.click(toggleBtn);
        const liveRegion = screen.getByText(/gallery (collapsed|expanded)/i);
        expect(liveRegion).toBeInTheDocument();
    });

    // Responsive/layout tests (simulate container width)
    it('renders correct grid span for mobile and desktop', () => {
        // Mock offsetWidth for mobile
        const refMock = {current: {offsetWidth: 350}};
        jest.spyOn(React, 'useRef').mockReturnValueOnce(refMock as any);
        renderWithProvider(<ImageGallery {...defaultProps} />);
        
        // Should render without error
        expect(screen.getByTestId('fileAttachmentList')).toBeInTheDocument();
    });
}); 