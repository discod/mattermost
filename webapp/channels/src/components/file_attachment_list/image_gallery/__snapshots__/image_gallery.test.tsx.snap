// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery matches snapshot 1`] = `
<div>
  <div
    class="image-gallery"
    data-testid="fileAttachmentList"
  >
    <div
      class="image-gallery__header"
    >
      <button
        aria-describedby="image-gallery-count"
        aria-expanded="true"
        class="image-gallery__toggle"
      >
        <svg
          fill="currentColor"
          height="16"
          version="1.1"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7,10L12,15L17,10H7Z"
          />
        </svg>
        <span
          id="image-gallery-count"
        >
          2 images
        </span>
      </button>
      <span
        class="image-gallery__separator"
      />
      <button
        class="image-gallery__download-all"
      >
        <svg
          fill="currentColor"
          height="12"
          version="1.1"
          viewBox="0 0 24 24"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19,18H5v2h14V18z M11,2h2v10.2l4.5-4.5l1.4,1.4L12,16L5.1,9.1l1.4-1.4l4.5,4.5V2z"
          />
        </svg>
        Download All
      </button>
    </div>
    <div
      class="image-gallery__body"
      role="list"
    >
      <div
        aria-label="Image 1 of 2"
        class="image-gallery__item image-gallery__item--small"
        role="listitem"
        tabindex="0"
      >
        <div
          class="file-view--single"
        >
          <div
            class="file__image"
          >
            <div
              class="image-header image-header--expanded"
            >
              <button
                aria-label="Toggle Embed Visibility"
                class="style--none single-image-view__toggle"
                data-expanded="true"
              >
                <svg
                  fill="currentColor"
                  height="16"
                  version="1.1"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7,10L12,15L17,10H7Z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="image-container"
              data-expanded="true"
              style="min-width: 50px; min-height: 50px;"
            >
              <div
                class="image-loaded"
              >
                <div
                  class="image-loading__container  single-image-view__image"
                  style="width: 100px; height: 100px; min-width: 50px; min-height: 50px;"
                >
                  <div
                    style="position: absolute; top: 50%; transform: translate(-50%, -50%); left: 50%;"
                  >
                    <div
                      class="file__image-loading"
                    >
                      <span
                        class="LoadingSpinner"
                        data-testid="loadingSpinner"
                        id="loadingSpinner"
                      >
                        <span
                          class="fa fa-spinner fa-fw fa-pulse spinner"
                          title="Loading Icon"
                        />
                      </span>
                      <span
                        class="loader-percent"
                      />
                    </div>
                  </div>
                  <svg
                    style="max-height: 100px; max-width: 100px; vertical-align: middle;"
                    viewBox="0 0 100 100"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
                <div
                  class="file-preview__button"
                  style="display: none;"
                >
                  <div
                    class="small-image__container"
                    style="min-width: 50px; min-height: 50px; display: flex; align-items: center; justify-content: center; max-width: 100%; max-height: 350px;"
                  >
                    <figure
                      class="image-loaded-container"
                      style="margin: 0px; display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
                    >
                      <img
                        aria-label="file thumbnail image1.png"
                        class=" single-image-view__image small-image--inside-container"
                        src="/api/v4/files/img1/preview"
                        style="object-fit: cover;"
                        tabindex="0"
                      />
                    </figure>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-label="Image 2 of 2"
        class="image-gallery__item image-gallery__item--small"
        role="listitem"
        tabindex="0"
      >
        <div
          class="file-view--single"
        >
          <div
            class="file__image"
          >
            <div
              class="image-header image-header--expanded"
            >
              <button
                aria-label="Toggle Embed Visibility"
                class="style--none single-image-view__toggle"
                data-expanded="true"
              >
                <svg
                  fill="currentColor"
                  height="16"
                  version="1.1"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7,10L12,15L17,10H7Z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="image-container"
              data-expanded="true"
              style="min-width: 50px; min-height: 50px;"
            >
              <div
                class="image-loaded"
              >
                <div
                  class="image-loading__container  single-image-view__image"
                  style="width: 300px; height: 200px; min-width: 50px; min-height: 50px;"
                >
                  <div
                    style="position: absolute; top: 50%; transform: translate(-50%, -50%); left: 50%;"
                  >
                    <div
                      class="file__image-loading"
                    >
                      <span
                        class="LoadingSpinner"
                        data-testid="loadingSpinner"
                        id="loadingSpinner"
                      >
                        <span
                          class="fa fa-spinner fa-fw fa-pulse spinner"
                          title="Loading Icon"
                        />
                      </span>
                      <span
                        class="loader-percent"
                      />
                    </div>
                  </div>
                  <svg
                    style="max-height: 200px; max-width: 300px; vertical-align: middle;"
                    viewBox="0 0 300 200"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
                <div
                  class="file-preview__button"
                  style="display: none;"
                >
                  <div
                    class="small-image__container"
                    style="min-width: 50px; min-height: 50px; display: flex; align-items: center; justify-content: center; max-width: 100%; max-height: 350px;"
                  >
                    <figure
                      class="image-loaded-container"
                      style="margin: 0px; display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
                    >
                      <img
                        aria-label="file thumbnail image2.jpg"
                        class=" single-image-view__image small-image--inside-container"
                        src="/api/v4/files/img2/preview"
                        style="object-fit: cover;"
                        tabindex="0"
                      />
                    </figure>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-live="polite"
      style="position: absolute; left: -9999px; height: 1px; width: 1px; overflow: hidden;"
    />
  </div>
</div>
`;
